### PR TITLE
Ghost, (minor) flying, and floor enemy update. Minor Player update. Flying enemy/homing enemy abstractification.

### PR DESCRIPTION
**Pull request by MattLiu-mygit**
Ghost enemy:
1. Now freezes when you look at it (or near it). 
         - Used some trig and big brain modifications(cuz gdscript coordinate system is wonky) to do 
            this :3. Pretty self-explanatory in the code hopefully.

2. Abstractified ghost enemy to inherit from the homing enemy AI type and moved it out of flying enemy for its own enemy AI base class.

Flying enemy/abstractification:
1. Helicopter enemy reverted to flying enemy to be a patrolling flying enemy base class.

Floor enemy:
1. Floor enemy bug causing it to glitch across gaps fixed. 
2. Floor enemy also doesn't cause knockback when attaching to player. 

Player:
1. Player modified to not be knocked back from hitboxes dealing 0 damage. 

Homing Enemy abstractification:
1. Implemented a homing base scene and script.

### CODE REVIEW

File: entities/enemies/Enemy.gd [hunk 1]
```diff
@@ -8,6 +8,10 @@ var motion = Vector2.ZERO
 onready var stats = $EnemyStats
 
 
+func ready() -> void:
+	set_physics_process(false)
```
Comment (JJamesWWang) [inline entities/enemies/Enemy.gd:12]: Just curious how this isn't causing enemies to freeze lol

File: entities/enemies/floor/FloorEnemy.gd [hunk 2]
```diff
@@ -1,20 +1,22 @@
 extends Enemy
 
-onready var original_floor_cast_position
+export(int) var SLOW_PERCENT = 0.25
```
Comment (JJamesWWang) [inline entities/enemies/floor/FloorEnemy.gd:3]: Slowing the player isn't universal to all floorenemies, just the stickyenemy, so move this to the stickyenemy script

File: entities/enemies/floor/FloorEnemy.gd [hunk 3]
```diff
@@ -1,20 +1,22 @@
...
+	# Prevents jumping off cliffs
+	if not floor_cast.is_colliding() and facing_direction == 1:
 		motion.x = 0
-	else:
-		floor_cast.position.x = facing_direction * original_floor_cast_position
+	elif not back_floor_cast.is_colliding() and facing_direction == -1:
+		motion.x = 0
+	elif facing_direction != 0: # and not attached_to_player:
+		floor_cast.position.x = facing_direction
 		motion.x = SPEED * facing_direction
```
Comment (JJamesWWang) [inline entities/enemies/floor/FloorEnemy.gd:20]: Hypothetically, if the Player flicked their mouse really quickly, could the floorenemy teleport to another floor?

File: entities/enemies/floor/StickyEnemy/StickyEnemy.gd [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+func _physics_process(delta) -> void:
+	var player = main_instances.player
+	
+	if attached_to_player and player != null:
+		motion.x = 0
+		player.motion.x *= SLOW_PERCENT
+		position.x = player.global_position.x
+		position.y = player.global_position.y + 12
+	
+	motion = move_and_slide(motion, Vector2.UP) * delta
```
Comment (JJamesWWang) [inline entities/enemies/floor/StickyEnemy/StickyEnemy.gd:16]: move_and_slide already uses delta so it shouldn't be multiplied here. Also, how is this moving if you're not calling ._physics_process(delta) in the base class?

File: entities/enemies/flying/FlyingEnemy.gd [hunk 5]
```diff
@@ -2,60 +2,69 @@ extends Enemy
...
+export(DIRECTION) var FLYING_DIRECTION = DIRECTION.RIGHT
 export(int) var LEFT_BOUND
 export(int) var RIGHT_BOUND
 export(int) var pause_time = 3
+export(float) var ACCELERATION = 0.75
 
 var start_x
 var motion_x
+var decelerating
+var e = 2.71828
```
Comment (JJamesWWang) [inline entities/enemies/flying/FlyingEnemy.gd:14]: This is only used in one method so you can change it to a local variable

File: entities/enemies/homing/ghost/GhostEnemy.gd [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+	var player = main_instances.player
+	
+	if spoopy_timer.time_left == 0:
+		spoopy_timer.start()
+	
+	if fade_timer.time_left == 0:
+		fade_timer.start()
+	
+	if player != null:
+		freeze_check(player)
```
Comment (JJamesWWang) [inline entities/enemies/homing/ghost/GhostEnemy.gd:34]: I'm confused how the ghostenemy works if you never call chase_player

File: entities/enemies/homing/ghost/GhostEnemy.gd [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		
+	# Freezes movement if ghost enemy within an angle range of where the 
+	# player looks
+	if ghost_angle < top_angle and ghost_angle > bottom_angle:
+		motion.x = 0
+		motion.y = 0
+
+
+func chase_player(player: KinematicBody2D, delta: float) -> void:
+	motion.y += spoopy_y_mod * spoopy_timer.time_left
```
Comment (JJamesWWang) [inline entities/enemies/homing/ghost/GhostEnemy.gd:51]: Also confused on how chase_player works when it's only one line

File: entities/enemies/Enemy.gd (hunk 1 above)
Comment (MattLiu-mygit) [reply entities/enemies/Enemy.gd:12]: I'm as baffled as you are xD

Comment (MattLiu-mygit) [reply entities/enemies/Enemy.gd:12]: Removed it though xD.

File: entities/enemies/floor/FloorEnemy.gd (hunk 3 above)
Comment (MattLiu-mygit) [reply entities/enemies/floor/FloorEnemy.gd:20]: No, now there are raycasts at either ends, so one of the raycasts will catch the floorenemy. Not modifying floor enemy raycast positions, so it isn't possible for them to move like that. I tested it with some gaps in the test rooms.

File: entities/enemies/floor/StickyEnemy/StickyEnemy.gd (hunk 4 above)
Comment (MattLiu-mygit) [reply entities/enemies/floor/StickyEnemy/StickyEnemy.gd:16]: Idk man, it just works xD. Though the base class should be fixed up now.

File: entities/enemies/homing/ghost/GhostEnemy.gd (hunk 6 above)
Comment (MattLiu-mygit) [reply entities/enemies/homing/ghost/GhostEnemy.gd:34]: chase_player was called in the base class.

File: entities/enemies/homing/ghost/GhostEnemy.gd (hunk 7 above)
Comment (MattLiu-mygit) [reply entities/enemies/homing/ghost/GhostEnemy.gd:51]: Check base class. It got abstractified. There are no overrides in gdscript. I'm just extending the chase_player from homing.gd in ghost.gd to fit the ghost movement.

File: entities/enemies/homing/ghost/GhostEnemy.gd (hunk 6 above)
Comment (JJamesWWang) [reply entities/enemies/homing/ghost/GhostEnemy.gd:34]: From my understanding of gdscript, if you override the base class, it won't be called by default. That's why I'm so confused.

File: entities/enemies/homing/ghost/GhostEnemy.gd (hunk 7 above)
Comment (JJamesWWang) [reply entities/enemies/homing/ghost/GhostEnemy.gd:51]: you have overridden the method with the same header? It's not an extension

**Pull request closed without merging**